### ISSUE TITLE
Not all Allocations are created when open ended Recurring Donations are updated

### ISSUE THREAD

**Issue by dospromptman** [SalesforceFoundation/Cumulus#1960]
When default allocations is enabled, there are recurring donations, and the forecast months setting is changed, then for each of the recurring donations there should be newly generated opportunities, payments, and allocations.  However, under some circumstances, not all of the allocations are being generated.

``` gherkin
Given Default Allocations are enabled
And Opportunity Forecast Months is set to 1
When I create 20 open ended recurring donations
And set the Opportunity Forecast Months to 12
Then there should be 240 Opportunity records related to the recurring donations
And there should be 240 Payment records related to the opportunities
And there should be 240 Allocation records related to the opportunities
And there should be 240 Allocation records related to the payments
```

The above scenario fails with only 200 Allocation records created, instead of the expected 240.

I have created a unit test to reproduce this bug: [RecurringDonationsAllocationLimitTest.cls](https://github.com/SalesforceFoundation/Cumulus/blob/recurring-donation-allocation-test/src/classes/RecurringDonationsAllocationLimitTest.cls)

**lurch: add

**Comment by LurchTheButler** [comment]
Tracking <a href="https://foundation.my.salesforce.com/a2x800000001yU3AAI" target="_blank">W-009399</a>

**Comment by LurchTheButler** [comment]
Tracking <a href="https://foundation.my.salesforce.com/a2x800000001yU8AAI" target="_blank">W-009400</a>

**Comment by chrismcmahon** [comment]
Wow, nice Cucumber-style description!

**Comment by kbromer** [comment]
The payments are created correctly though, right?

**Comment by dospromptman** [comment]
I only tested the number of payments created, which was 240 as expected.  I did not test other properties of the created payments.

**Comment by dospromptman** [comment]
This issue seems to be related to recursion control flags set in `ALLO_Allocations_TDTM` that are not being reset between executions.

**Comment by davidhabib** [comment]
depending on how your test code is structured, it is the responsibility of the test code to reset flags that normally would have been reset by the context expiring.  in other words, if you do a bunch of separate DML operations in a row, which could not normally occur in the ui, then they all execute in one context when normally the context would have expired (and all static variables reset).  Thus you'll find various places in test code that we have to manually reset static flags.  This is very common after initially creating test data, before running the actual testing logic.

**Comment by dospromptman** [comment]
Yes, I was thinking that the fact that it is running in a test might be contributing to the issue demonstrated, since from the debug logs I can see that the entire test runs within a single execution context.  I would have thought that the static recursion flags would have gotten reset during subsequent chunks of the trigger executing, in actual usage.

 However, from those debug logs I can also see that even though the entire thing seems to take place within a single execution context, the variables DO seem to get reset at some points.  I was unable to locate the code that might be resetting them.

 My next step is to attempt to verify this outside of the context of a unit test.

**Comment by njjc** [comment]
I think @davidhabib is right. Since the recurring donations batch job is doing a dml of 200, then 40, and the allocations trigger uses a static flag for trigger execution, the flag is set on the second batch. This isn't the case outside of tests since each batch is in a separate context.

**Comment by judisohn** [comment]
See #2591 this should probably close when that does.

**Comment by mrbelvedere** [comment]
Included in beta release 3.104 (Beta 3)
